glasbey.py

```diff
@@ -38,118 +45,227 @@ def finish(self):
-        palette.append((rgb[0] * 256 + rgb[1]) * 256 + rgb[2])
-    return palette
-
-
-def save_palette(palette, f, fmt):
-    if fmt == "byte":
-        for color in palette:
-            rgb255 = tuple(int(round(k * 255)) for k in color)
-            f.write("{},{},{}\n".format(*rgb255))
-    else:
+            assert not self.overwrite_base_palette, "no base_palette specified, cannot overwrite it!"
+
+        # Load colors
+        self.colors = self.load_colors()
+
+        # Initialize base palette
+        if type(base_palette) == str:
+            self.base_palette = base_palette
+            self.palette = self.load_palette(base_palette)
+            self.palette = [self.colors[i, :] for i in self.palette]
+        elif type(base_palette) == list and len(base_palette) > 0:
+            self.palette = [(rgb[0] * 256 + rgb[1]) * 256 + rgb[2] for rgb in base_palette]
+            self.palette = [self.colors[i, :] for i in self.palette]
+        else:
+            self.palette = [self.colors[-1, :]]  # white
+
+        assert self.palette_invariant(), "Internal error during __init__: self.palette is poorly formatted."
+
+        # Update self.colors
+        # Exclude greys (values with low Chroma in JCh) and set lightness range,
+        if lightness_range is not None:
+            jch = cspace_convert(self.colors, "CAM02-UCS", "JCh")
+            self.colors = self.colors[
+                          (jch[:, 0] >= lightness_range[0]) & (jch[:, 0] <= lightness_range[1]), :
+                          ]
+        if chroma_range is not None:
+            jch = cspace_convert(self.colors, "CAM02-UCS", "JCh")
+            self.colors = self.colors[
+                          (jch[:, 1] >= chroma_range[0]) & (jch[:, 1] <= chroma_range[1]), :
+                          ]
+        if hue_range is not None:
+            jch = cspace_convert(self.colors, "CAM02-UCS", "JCh")
+            if hue_range[0] > hue_range[1]:
+                self.colors = self.colors[
+                              (jch[:, 2] >= hue_range[0]) | (jch[:, 2] <= hue_range[1]), :
+                              ]
+            else:
+                self.colors = self.colors[
+                              (jch[:, 2] >= hue_range[0]) & (jch[:, 2] <= hue_range[1]), :
+                              ]
+        # Exclude colors that are close to black
+        if no_black:
+            MIN_DISTANCE_TO_BLACK = 35
+            d = np.linalg.norm((self.colors - self.colors[0, :]), axis=1)
+            self.colors = self.colors[d > MIN_DISTANCE_TO_BLACK, :]
+
+    def get_palette(self, size):
+        """
+        Return palette in sRGB1 format.
+
+        If the palette isn't long enough, new entries are generated.
+        """
+
+        if size > len(self.palette):
+            # need to extend palette
+            self.extend_palette(size=size)
+        return cspace_convert(self.palette[0:size], "CAM02-UCS", "sRGB1")
+
+    def load_colors(self):
```

Comment:
I'd call this `load_or_generate_color_table()`.

Reply:
Done.